compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -510,6 +512,21 @@ impl<'tcx> ObligationCauseCode<'tcx> {
         base_cause
     }
 
+    /// Returns the base obligation and the base trait predicate, if any, ignoring
+    /// derived obligations.
+    pub fn peel_derives_with_predicate(&self) -> (&Self, Option<ty::PolyTraitPredicate<'tcx>>) {
+        let mut base_cause = self;
+        let mut base_trait_pred = None;
+        while let Some((parent_code, parent_pred)) = base_cause.parent() {
+            base_cause = parent_code;
+            if let Some(parent_pred) = parent_pred {
```

Comment:
This will return predicates that don't actually line up with the derived cause if `parent_pred` is `None`. This is partly why #119352 is failing.

Reply:
https://github.com/rust-lang/rust/pull/119361#pullrequestreview-1797517286